jsonargparse_tests/test_cli.py

```diff
@@ -120,6 +121,31 @@ def test_multiple_functions_subcommand_help():
     assert "--a2 A2" in out
 
 
+def conditionalA(foo: int = 1):
+    return foo
+
+
+def conditionalB(bar: int = 2):
+    return bar
+
+
+def conditional_function(fn: "Literal['A', 'B']", *args, **kwargs):
+    if fn == "A":
+        return conditionalA(*args, **kwargs)
+    elif fn == "B":
+        return conditionalB(*args, **kwargs)
+    raise NotImplementedError(fn)
+
+
+@pytest.mark.skipif(condition=sys.version_info < (3, 9), reason="python>=3.9 is required")
+@pytest.mark.skipif(condition=not Literal, reason="Literal is required")
+def test_literal_conditional_function():
+    out = get_cli_stdout(conditional_function, args=["--help"])
+    assert "Conditional arguments" in out
+    assert "--foo FOO             (type: int, default: Conditional<ast-resolver> {1, NOT_ACCEPTED})" in out
+    assert "--bar BAR             (type: int, default: Conditional<ast-resolver> {2, NOT_ACCEPTED})" in out
```

Comment:
The repro in the issue works, but in my real script, these conditional instances get actually passed down to the selected function, causing an error:

```python
ValueError: Expected "default" argument to be an instance of "TrainArgs" or its kwargs dict, given Conditional<ast-resolver> {TrainArgs(save_interval=1000, log_interval=1, global_batch_size=128, micro_batch_size=4, lr_warmup_steps=100, epochs=5, max_tokens=None, max_steps=None, max_seq_length=None, tie_embeddings=None, learning_rate=0.0003, weight_decay=0.02, beta1=0.9, beta2=0.95, max_norm=None, min_lr=6e-05), TrainArgs(save_interval=1000, log_interval=1, global_batch_size=64, micro_batch_size=1, lr_warmup_steps=100, epochs=5, max_tokens=None, max_steps=None, max_seq_length=None, tie_embeddings=None, learning_rate=0.003, weight_decay=0.02, beta1=0.9, beta2=0.95, max_norm=None, min_lr=6e-05), TrainArgs(save_interval=1000, log_interval=1, global_batch_size=64, micro_batch_size=4, lr_warmup_steps=100, epochs=5, max_tokens=None, max_steps=None, max_seq_length=None, tie_embeddings=None, learning_rate=0.001, weight_decay=0.02, beta1=0.9, beta2=0.95, max_norm=None, min_lr=6e-05), TrainArgs(save_interval=1000, log_interval=1, global_batch_size=128, micro_batch_size=2, lr_warmup_steps=100, epochs=5, max_tokens=None, max_steps=None, max_seq_length=None, tie_embeddings=None, learning_rate=0.001, weight_decay=0.02, beta1=0.9, beta2=0.95, max_norm=None, min_lr=6e-05)}
```

`TrainArgs` is a dataclass, with the argument set as:

`train: TrainArgs = TrainArgs()`

Reply:
Is this code available somewhere so that I can reproduce this and understand what is going on?

Reply:
The error above would be something to fix in a different pull request. For now I will merge this and look at that problem later. If you like, create an issue with just the same comment as here.